storage-proofs/src/post/fallback/vanilla.rs

```diff
@@ -0,0 +1,571 @@
+            .map(MerkleProof::root)
+            .collect()
+    }
+
+    #[allow(clippy::type_complexity)]
+    pub fn paths(&self) -> Vec<&Vec<(Vec<H::Domain>, usize)>> {
+        self.inclusion_proofs
+            .iter()
+            .map(MerkleProof::path)
+            .collect()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct FallbackPoSt<'a, H>
+where
+    H: 'a + Hasher,
+{
+    _h: PhantomData<&'a H>,
+}
+
+pub fn generate_sector_challenges<T: Domain>(
+    randomness: T,
+    challenge_count: usize,
+    sectors: &OrderedSectorSet,
+) -> Result<Vec<SectorId>> {
+    (0..challenge_count)
+        .map(|n| generate_sector_challenge(randomness, n, sectors))
+        .collect()
+}
+
+/// Generate a single sector challenge.
+pub fn generate_sector_challenge<T: Domain>(
+    randomness: T,
+    n: usize,
+    sectors: &OrderedSectorSet,
+) -> Result<SectorId> {
+    let mut hasher = Sha256::new();
+    hasher.input(AsRef::<[u8]>::as_ref(&randomness));
+    hasher.input(&n.to_le_bytes()[..]);
+
+    let hash = hasher.result();
+
+    let sector_challenge = LittleEndian::read_u64(&hash.as_ref()[..8]);
+    let sector_index = (sector_challenge % sectors.len() as u64) as usize;
+    let sector = *sectors
+        .iter()
+        .nth(sector_index)
+        .context("invalid challenge generated")?;
+
+    Ok(sector)
+}
+
+/// Generate all challenged leaf ranges for a single sector, such that the range fits into the sector.
+pub fn generate_leaf_challenges<T: Domain>(
+    pub_params: &PublicParams,
+    randomness: T,
+    sector_id: u64,
+    challenge_count: usize,
+) -> Result<Vec<u64>> {
+    let mut challenges = Vec::with_capacity(challenge_count);
+
+    for leaf_challenge_index in 0..challenge_count {
+        let challenge = generate_leaf_challenge(
+            pub_params,
+            randomness,
+            sector_id,
+            leaf_challenge_index as u64,
+        )?;
+        challenges.push(challenge)
+    }
+
+    Ok(challenges)
+}
+
+/// Generates challenge, such that the range fits into the sector.
+pub fn generate_leaf_challenge<T: Domain>(
+    pub_params: &PublicParams,
+    randomness: T,
+    sector_id: u64,
+    leaf_challenge_index: u64,
+) -> Result<u64> {
+    let mut hasher = Sha256::new();
+    hasher.input(AsRef::<[u8]>::as_ref(&randomness));
+    hasher.input(&sector_id.to_le_bytes()[..]);
+    hasher.input(&leaf_challenge_index.to_le_bytes()[..]);
+    let hash = hasher.result();
+
+    let leaf_challenge = LittleEndian::read_u64(&hash.as_ref()[..8]);
+
+    let challenged_range_index = leaf_challenge % (pub_params.sector_size / NODE_SIZE as u64);
+
+    Ok(challenged_range_index)
+}
+
+impl<'a, H: 'a + Hasher> ProofScheme<'a> for FallbackPoSt<'a, H> {
+    type PublicParams = PublicParams;
+    type SetupParams = SetupParams;
+    type PublicInputs = PublicInputs<'a, H::Domain>;
+    type PrivateInputs = PrivateInputs<'a, H>;
+    type Proof = Proof<H>;
+    type Requirements = NoRequirements;
```

Comment:
I think this should actually be `ChallengeRequirements` not `NoRequirements`. Similar to PoRep, we need to ensure that a set of provided PoSt circuit proofs actually contains enough proofs.

I know that is being checked in a different way here in the vanilla proof, but we need to do it also in the circuit `MultiProof` case, and that is the intended purpose of `ChallengeRequirements`.

Reply:
fixed, though I am not sure it does verify everything we need,

Reply:
I don't think you updated `POREP_MINIMUM_CHALLENGES` in `constants.rs` can you confirm that 138 is the correct number of minimum challenges for 32GiB? (cc: @nicola)

What else do you think needs to be verified here that might be missing? Let's think it through and extend the relevant implementation of `ProofScheme::satisfies_requirements` (and `ChallengeRequirements`, if necessary) to ensure everything necessary *is* verified.

Reply:
UPDATE 2: I see now that you did implement `satisfies_requirements`, but you must not ignore the number of partitions in the input — since that is what needs to be verified.

Reply:
updated again